File: penguin-popper/orig/imgGraphic.java
public class imgGraphic{
	java.awt.Image[][] img;
	double start, end;
	int size, frames;
	public imgGraphic (int z, double st, double en, int fr, java.awt.Color[] colors, double downShift, String s){
		size=z;
		start=st;
		end=en;
		frames=fr;
		img=new java.awt.Image[colors.length][frames+1];
		for (int color=0; color<colors.length; color++){
			for (int frame=0; frame<frames+1; frame++){
				java.awt.image.BufferedImage temp=new java.awt.image.BufferedImage(100, 100, java.awt.image.BufferedImage.TYPE_INT_ARGB);
				java.awt.Graphics2D g2=(java.awt.Graphics2D)temp.getGraphics();
				g2.setRenderingHint(java.awt.RenderingHints.KEY_ANTIALIASING, java.awt.RenderingHints.VALUE_ANTIALIAS_ON);
				new graphic(s).draw(g2, 50, 50*(1+downShift), 100, start+frame*(end-start)/frames, false, colors[color]);
				img[color][frame]=temp.getScaledInstance(size, size, java.awt.image.BufferedImage.SCALE_SMOOTH);
			}
		}
	}
	// angle=a1+(a2-a1)/anglesPerRadian*index
	// (angle-a1)/(a2-a1)*anglesPerRadian=index
	public void draw(java.awt.Graphics2D g2, double xc, double yc, double angle, int vColor){
		g2.drawImage(img[vColor][(int)((angle%(2*Math.PI)-start)/(end-start)*frames)], (int)Math.round(xc-size/2), (int)Math.round(yc-size/2), null);
	}
}

File: penguin-popper/imgGraphic.js
import {graphic} from './graphic.js';

export class imgGraphic {
  constructor(z, st, en, fr, colors, downShift, s) {
    this.size = z;
    this.start = st;
    this.end = en;
    this.frames = fr;
    this.img = [];
    for (let color = 0; color < colors.length; color++) {
      this.img[color] = [];
      for (let frame = 0; frame < this.frames + 1; frame++) {
        let canvas = document.createElement('canvas');
        canvas.width = canvas.height = 100;
        let ctx = canvas.getContext('2d');
        new graphic(s).draw(
          ctx,
          50,
          50 * (1 + downShift),
          50,
          this.start + (frame * (this.end - this.start)) / this.frames,
          false,
          colors[color]
        );
        this.img[color][frame] = canvas;
        // document.body.append(canvas);
      }
    }
  }
  draw(ctx, xc, yc, angle, vColor) {
    ctx.drawImage(
      this.img[Math.floor(vColor)][
        Math.floor(
          (((angle % (2 * Math.PI)) - this.start) / (this.end - this.start)) *
            this.frames
        )
      ],
      Math.round(xc - this.size / 2),
      Math.round(yc - this.size / 2)
    );
  }
}
